File: backend/src/main/java/cn/etarch/mao/agent/service/AgentService.java
package cn.etarch.mao.agent.service;

import cn.etarch.mao.agent.entity.Agent;
import cn.etarch.mao.agent.entity.AgentExperience;
import cn.etarch.mao.agent.entity.AgentTag;
import cn.etarch.mao.agent.mapper.AgentMapper;
import cn.etarch.mao.agent.mapper.AgentTagMapper;
import cn.etarch.mao.common.exception.BusinessException;
import cn.etarch.mao.common.result.ErrorCode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class AgentService {

    private final AgentMapper agentMapper;
    private final AgentTagMapper agentTagMapper;
    private final AgentExperienceService experienceService;
    private final ObjectMapper objectMapper;

    public List<Agent> listAgents(Long userId, String keyword) {
        QueryWrapper<Agent> qw = new QueryWrapper<>();
        if (keyword != null && !keyword.isEmpty()) {
            qw.like("name", keyword);
        }
        qw.orderByDesc("is_default").orderByDesc("created_at");
        return agentMapper.selectList(qw);
    }

    public Agent getAgent(Long id) {
        Agent agent = agentMapper.selectById(id);
        if (agent == null) {
            throw new BusinessException(ErrorCode.AGENT_NOT_FOUND);
        }
        return agent;
    }

    public Agent getDefaultAgent() {
        return agentMapper.selectOne(new QueryWrapper<Agent>().eq("is_default", 1).last("LIMIT 1"));
    }

    public Agent requireDefaultAgent() {
        Agent agent = getDefaultAgent();
        if (agent == null) {
            throw new BusinessException(ErrorCode.AGENT_NOT_FOUND, "未配置默认 Agent，请先在管理后台设置");
        }
        return agent;
    }

    @Transactional
    public Agent createAgent(Long userId, String name, String description,
                              String systemPrompt,
                              List<String> tags,
                              List<String> skillNames,
                              List<AgentExperienceService.ExperienceInput> experiences,
                              Integer isDefault) {
        if (isDefault != null && isDefault == 1) {
            clearDefaultFlag();
        }
        Agent agent = new Agent();
        agent.setName(name);
        agent.setDescription(description);
        agent.setSystemPrompt(systemPrompt);
        agent.setCreatorId(userId);
        agent.setIsDefault(isDefault != null ? isDefault : 0);
        if (skillNames != null && !skillNames.isEmpty()) {
            try {
                agent.setSkillNames(objectMapper.writeValueAsString(skillNames));
            } catch (Exception e) {
                // ignore serialization error
            }
        }
        agentMapper.insert(agent);

        // Save tags
        if (tags != null) {
            for (String tag : tags) {
                AgentTag agentTag = new AgentTag();
                agentTag.setAgentId(agent.getId());
                agentTag.setTag(tag);
                agentTagMapper.insert(agentTag);
            }
        }

        if (experiences != null) {
            experienceService.syncExperiences(agent.getId(), experiences);
        }

        return agent;
    }

    @Transactional
    public Agent updateAgent(Long id, String name, String description,
                              String systemPrompt,
                              List<String> skillNames,
                              List<String> tags,
                              List<AgentExperienceService.ExperienceInput> experiences,
                              Integer isDefault) {
        Agent agent = getAgent(id);
        if (name != null) agent.setName(name);
        if (description != null) agent.setDescription(description);
        if (systemPrompt != null) agent.setSystemPrompt(systemPrompt);
        if (skillNames != null) {
            try {
                agent.setSkillNames(skillNames.isEmpty() ? null : objectMapper.writeValueAsString(skillNames));
            } catch (Exception e) {
                // ignore serialization error
            }
        }
        if (isDefault != null) {
            if (isDefault == 1) {
                clearDefaultFlag();
            }
            agent.setIsDefault(isDefault);
        }
        agentMapper.updateById(agent);

        // Update tags (delete old + insert new)
        if (tags != null) {
            agentTagMapper.delete(new QueryWrapper<AgentTag>().eq("agent_id", id));
            for (String tag : tags) {
                AgentTag agentTag = new AgentTag();
                agentTag.setAgentId(id);
                agentTag.setTag(tag);
                agentTagMapper.insert(agentTag);
            }
        }

        // experiences == null → 不改动；[] → 清空
        experienceService.syncExperiences(id, experiences);

        return agent;
    }

    @Transactional
    public void deleteAgent(Long id) {
        Agent agent = getAgent(id);
        if (agent.getIsDefault() != null && agent.getIsDefault() == 1) {
            throw new BusinessException(ErrorCode.AGENT_IS_DEFAULT);
        }
        agentTagMapper.delete(new QueryWrapper<AgentTag>().eq("agent_id", id));
        experienceService.deleteByAgentId(id);
        agentMapper.deleteById(id);
    }

    public List<AgentTag> getAgentTags(Long agentId) {
        return agentTagMapper.selectList(
                new QueryWrapper<AgentTag>().eq("agent_id", agentId));
    }

    public List<AgentExperience> getAgentExperiences(Long agentId) {
        return experienceService.listByAgentId(agentId);
    }

    private void clearDefaultFlag() {
        Agent update = new Agent();
        update.setIsDefault(0);
        agentMapper.update(update, new QueryWrapper<Agent>().eq("is_default", 1));
    }
}

File: backend-ts/src/agent/agent.service.ts
import { validateAgentAvatarUrl } from './agent-avatar.js';
import { BusinessException } from '../common/business-exception.js';
import { ErrorCode } from '../common/error-code.js';
import type { AgentExperienceService } from './agent-experience.service.js';
import type { AgentSuggestedQuestionService } from './agent-suggested-question.service.js';
import type {
  Agent,
  AgentExperience,
  AgentRepository,
  AgentSuggestedQuestion,
  ExperienceInput,
  SuggestedQuestionInput,
} from './types.js';

/** Agent 默认模型校验所需的最小模型查询能力 */
export interface AgentModelLookup {
  findById(id: number): Promise<{ id?: number; status?: number | null } | null>;
}

export class AgentService {
  constructor(
    private readonly agentRepo: AgentRepository,
    private readonly experienceService: AgentExperienceService,
    private readonly suggestedQuestionService: AgentSuggestedQuestionService,
    private readonly modelLookup?: AgentModelLookup,
  ) {}

  listAgents(_userId: number, keyword?: string | null, includeDisabled = false): Promise<Agent[]> {
    return this.agentRepo.selectList(keyword, includeDisabled);
  }

  async getAgent(id: number): Promise<Agent> {
    const agent = await this.agentRepo.findById(id);
    if (!agent) {
      throw new BusinessException(ErrorCode.AGENT_NOT_FOUND);
    }
    return agent;
  }

  getDefaultAgent(): Promise<Agent | null> {
    return this.agentRepo.findDefault();
  }

  async requireDefaultAgent(): Promise<Agent> {
    const agent = await this.getDefaultAgent();
    if (!agent) {
      throw new BusinessException(ErrorCode.AGENT_NOT_FOUND, '未配置默认 Agent，请先在管理后台设置');
    }
    return agent;
  }

  async createAgent(
    userId: number,
    name: string,
    description: string | null | undefined,
    systemPrompt: string,
    skillNames: string[] | null | undefined,
    mcpServerIds: number[] | null | undefined,
    experiences: ExperienceInput[] | null | undefined,
    suggestedQuestions: SuggestedQuestionInput[] | null | undefined,
    isDefault: number | null | undefined,
    defaultModelId: number | null | undefined,
    avatarUrl?: string | null,
  ): Promise<Agent> {
    validateAgentAvatarUrl(avatarUrl);
    if (isDefault != null && isDefault === 1) {
      await this.agentRepo.clearDefaultFlag();
    }
    await this.requireEnabledModel(defaultModelId);
    const agent: Agent = {
      name,
      description,
      avatarUrl: avatarUrl ?? null,
      systemPrompt,
      creatorId: userId,
      defaultModelId: defaultModelId ?? null,
      isDefault: isDefault != null ? isDefault : 0,
      enabled: 1,
    };
    if (skillNames != null && skillNames.length > 0) {
      agent.skillNames = JSON.stringify(skillNames);
    }
    if (mcpServerIds != null && mcpServerIds.length > 0) {
      agent.mcpServerIds = JSON.stringify(mcpServerIds);
    }
    await this.agentRepo.insert(agent);

    if (experiences != null) {
      await this.experienceService.syncExperiences(agent.id!, experiences);
    }
    if (suggestedQuestions != null) {
      await this.suggestedQuestionService.syncSuggestedQuestions(agent.id!, suggestedQuestions);
    }

    return agent;
  }

  async updateAgent(
    operatorId: number,
    id: number,
    name: string | null | undefined,
    description: string | null | undefined,
    systemPrompt: string | null | undefined,
    skillNames: string[] | null | undefined,
    mcpServerIds: number[] | null | undefined,
    experiences: ExperienceInput[] | null | undefined,
    suggestedQuestions: SuggestedQuestionInput[] | null | undefined,
    isDefault: number | null | undefined,
    defaultModelId: number | null | undefined,
    avatarUrl?: string | null,
  ): Promise<Agent> {
    validateAgentAvatarUrl(avatarUrl);
    const agent = await this.getAgent(id);
    if (avatarUrl !== undefined) agent.avatarUrl = avatarUrl;
    if (name != null) agent.name = name;
    if (description != null) agent.description = description;
    if (systemPrompt != null) agent.systemPrompt = systemPrompt;
    if (skillNames != null) {
      agent.skillNames = skillNames.length === 0 ? null : JSON.stringify(skillNames);
    }
    if (mcpServerIds != null) {
      agent.mcpServerIds = mcpServerIds.length === 0 ? null : JSON.stringify(mcpServerIds);
    }
    if (defaultModelId !== undefined) {
      await this.requireEnabledModel(defaultModelId);
      agent.defaultModelId = defaultModelId;
    }
    if (isDefault != null) {
      if (isDefault === 1) {
        if (agent.enabled === 0) {
          throw new BusinessException(ErrorCode.PARAM_INVALID, '已停用的 Agent 不能设为默认，请先启用');
        }
        await this.agentRepo.clearDefaultFlag();
      }
      agent.isDefault = isDefault;
    }
    await this.agentRepo.updateById(agent, operatorId, systemPrompt != null);

    await this.experienceService.syncExperiences(id, experiences);
    await this.suggestedQuestionService.syncSuggestedQuestions(id, suggestedQuestions);
    return agent;
  }

  async listPromptVersions(id: number) {
    await this.getAgent(id);
    return this.agentRepo.listPromptVersions(id);
  }

  async rollbackPrompt(id: number, version: number, operatorId: number): Promise<Agent> {
    if (!Number.isSafeInteger(version) || version <= 0) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '版本号必须为正整数');
    }
    return this.agentRepo.rollbackPrompt(id, version, operatorId);
  }

  async setEnabled(id: number, enabled: boolean): Promise<Agent> {
    const agent = await this.getAgent(id);
    const next = enabled ? 1 : 0;
    if (next === 0 && agent.isDefault === 1) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '不能停用默认 Agent，请先指定其他默认 Agent');
    }
    if ((agent.enabled ?? 1) !== next) {
      await this.agentRepo.updateEnabled(id, next);
      agent.enabled = next;
    }
    return agent;
  }

  async deleteAgent(id: number): Promise<void> {
    const agent = await this.getAgent(id);
    if (agent.isDefault != null && agent.isDefault === 1) {
      throw new BusinessException(ErrorCode.AGENT_IS_DEFAULT);
    }
    await this.experienceService.deleteByAgentId(id);
    await this.suggestedQuestionService.deleteByAgentId(id);
    await this.agentRepo.deleteById(id);
  }

  /** 默认模型必须存在且启用；null/undefined 表示未配置，直接跳过 */
  private async requireEnabledModel(defaultModelId: number | null | undefined): Promise<void> {
    if (defaultModelId == null) return;
    if (this.modelLookup == null) return;
    const model = await this.modelLookup.findById(defaultModelId);
    if (model == null || model.status !== 1) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '默认模型不存在或未启用');
    }
  }

  async removeSkillNameFromAll(skillName: string): Promise<number> {
    return this.agentRepo.removeSkillName(skillName);
  }

  getAgentExperiences(agentId: number): Promise<AgentExperience[]> {
    return this.experienceService.listByAgentId(agentId);
  }

  getAgentSuggestedQuestions(agentId: number): Promise<AgentSuggestedQuestion[]> {
    return this.suggestedQuestionService.listByAgentId(agentId);
  }
}
